mittab/libs/assign_rooms.py

```diff
@@ -0,0 +1,65 @@
+import random
+from mittab.apps.tab.models import RoomCheckIn, Round, TabSettings
+from mittab.libs import errors, mwmatching, tab_logic
+from django.db import transaction
+
+
+def add_rooms():
+    no_seeding = TabSettings.get("disable_room_seeding", 0)
+    
+    #Clear any existing room assignments
+    Round.objects.filter(round_number=TabSettings.get("cur_round") - 1).update(room=None)
+    round_number = TabSettings.get("cur_round") - 1
+
+    rooms = RoomCheckIn.objects.filter(round_number=round_number).select_related("room")
+    rooms = sorted((r.room for r in rooms), key=lambda r: r.rank, reverse=True)
+    pairings = tab_logic.sorted_pairings(round_number)
+
+    if no_seeding:
+        random.shuffle(pairings)
+
+    if not pairings or not rooms or len(pairings) > len(rooms):
+        raise errors.RoomAssignmentError("Not enough rooms or pairings")
+
+    graph_edges = []
+    
+    for pairing_i, pairing in enumerate(pairings):
+        for room_i, room in enumerate(rooms):
+            # I chose to put this logic directly into the loop because some later changes that aren't
+            # in this branch require some data prep that made breaking the weight calculation out
+            # a lot less natural, and this ended up being a bit more readable
+            weight = 0
+
+            #High seed high room bonus
+            if no_seeding == 0:
+                weight -= abs(pairing_i - room_i)
+
+            #Bad room penalty
+            weight -= room.rank
+
+            edge = (pairing_i, len(pairings) + room_i, weight)
+            graph_edges.append(edge)
```

Comment:
This sort of match alg isn't intuitive to me, but I can see how it could work. Would the idea be to score the matching based on the specific pairing it's being assigned to?

e.g. With these two pairings:
 - Top-seeded round with debater(s)/judge with a wheel chair
 - Bottom-seeded team with no accessibility requirements

And these rooms:
 - Room in GA building, but not wheelchair accessible
 - Room a few minutes away, but is wheelchair accessible

We'd want the weight of the matching to be dynamic, where the wheelchair accessible room is the top match for the first round but not the second

Reply:
Yeah, pretty much, especially for the "high seed high room" bonus, which replicates the existing functionality of trying to give the n-ranked room to the n-ranked round.

For tags, the idea isn't necessarily that accessible rooms will get bad weights for debaters who don't need them as much as just putting a substantial negative weight on unfulfilled room tags, meaning max weight matching will first do whatever it can to fulfill all room tags (and to fulfill the higher weight ones first), and only then look at the smaller priorities like prioritizing good rooms, and giving better teams better rooms.

This is roughly what that will look like:
```
            #Unfilled tags penalty
            room_tags = room_tags_dict[room.id]
            unfulfilled_tags = pairing_tags - room_tags
            if unfulfilled_tags:
                unfulfilled_tags_matrix[pairing_i][room_i] = unfulfilled_tags

                #Arbitrary weight that's large enough to gaurentee precedence
                weight -= 1000 * sum(tag.priority for tag in unfulfilled_tags) 
```